src/transformers/modeling_utils.py

```diff
@@ -778,50 +789,84 @@ def generate(
             pad_token_id = eos_token_ids[0]
 
         # current position and vocab size
-        cur_len = input_ids.shape[1]
         vocab_size = self.config.vocab_size
 
-        if num_return_sequences != 1 and do_sample:
-            # Expand input to num return sequences
-            input_ids = input_ids.unsqueeze(1).expand(batch_size, num_return_sequences, cur_len)
-            input_ids = input_ids.contiguous().view(
-                batch_size * num_return_sequences, cur_len
-            )  # shape: (batch_size * num_return_sequences, cur_len)
+        # set effective batch size and effective batch multiplier according to do_sample
+        if do_sample:
             effective_batch_size = batch_size * num_return_sequences
+            effective_batch_mult = num_return_sequences
         else:
             effective_batch_size = batch_size
+            effective_batch_mult = 1
+
+        # Expand input ids if num_beams > 1 or num_return_sequences > 1
+        if num_return_sequences > 1 or num_beams > 1:
+            input_ids_len = input_ids.shape[-1]
+            input_ids = input_ids.unsqueeze(1).expand(batch_size, effective_batch_mult * num_beams, input_ids_len)
+            input_ids = input_ids.contiguous().view(
+                effective_batch_size * num_beams, input_ids_len
+            )  # shape: (batch_size * num_return_sequences * num_beams, cur_len)
+
+        # TODO (PVP): probably not the best way to check whether model is encoder decoder
+        is_encoder_decoder = (
+            hasattr(self, "model") and hasattr(self.model, "decoder") and hasattr(self.model, "encoder")
+        )
+        if is_encoder_decoder:
+            eos_token_id = eos_token_ids[0]
+            assert bos_token_id is not None, "Encoder Decoder Models need to have a bos_token_id"
+            assert eos_token_id is not None, "Encoder Decoder Models need to have a eos_token_id"
+            # encoder decoder need to start with empty input_ids and copy the input_ids to encoder_inputs
```

Comment:
Regarding 3). 
Not really sure the decoder_input_ids were initialized with the <EOS> token - is this a hack @sshleifer 
For now seems to also work with <BOS> token initialization which is the better way to do it I think.

Reply:
I don't understand "the token" and "token initialization" in your comment, could you clarify?

Reply:
sorry, just updated it